altair/utils/core.py

```diff
@@ -41,9 +50,21 @@
     from altair.utils._dfi_types import DataFrame as DfiDataFrame
     from altair.vegalite.v5.schema._typing import StandardType_T as InferredVegaLiteType
 
-V = TypeVar("V")
-P = ParamSpec("P")
 TIntoDataFrame = TypeVar("TIntoDataFrame", bound=IntoDataFrame)
+T = TypeVar("T")
+P = ParamSpec("P")
+R = TypeVar("R")
+
+WrapsFunc = TypeAliasType("WrapsFunc", Callable[..., R], type_params=(R,))
```

Comment:
Could you elaborate on the benefit of using `TypeAliasType` instead of `WrapsFunc: TypeAlias = Callable[..., R]`? I can't find this way of using it in https://docs.python.org/3/library/typing.html#type-aliases which either recommends `type WrapsFunc = Callable[..., R]` for Python 3.12 (hence we of course can't use it) or the `TypeAlias` way.

Reply:
Apologies if I end up repeating things you've already understood. 

https://docs.python.org/3/library/typing.html#type-aliases states

> A type alias is defined using the [`type`](https://docs.python.org/3/reference/simple_stmts.html#type) statement, which creates an instance of [`TypeAliasType`](https://docs.python.org/3/library/typing.html#typing.TypeAliasType). In this example, `Vector` and `list[float]` will be treated equivalently by static type checkers:

The benefit of using `TypeAliasType` is to have a [**generic**](https://typing.readthedocs.io/en/latest/spec/aliases.html#type-statement) alias (over the `type_params` parameter).
I think the doc for [`typing_extensions.TypeAliasType`](https://github.com/python/typing_extensions/blob/d9509f902c20e3d51c62b8abe522809b4760c0ff/src/typing_extensions.py#L3455-L3481) is a bit more useful to demonstrate this. 

I've used this for [`OneOrSeq`](https://github.com/vega/altair/blob/5207768b6e533c0509218376942309d1c7bac22f/altair/vegalite/v5/schema/_typing.py#L73-L87) as well, which let us specialize its `type_params` when defining [`channels.EncodeKwds`](https://github.com/vega/altair/blob/5207768b6e533c0509218376942309d1c7bac22f/altair/vegalite/v5/schema/channels.py#L25645).

[PEP 695](https://peps.python.org/pep-0695/#generic-type-alias) also has some examples of generic aliases - it is that syntax which this functionality is replicating - but in a backwards-compatible way

Reply:
Maybe I'm missing something but defining a type alias the "old" way also produces generic aliases, see [mypy docs]. Or here a small demo:

```python
from typing import List, Tuple, Union
from typing_extensions import TypeAlias, TypeAliasType, TypeVar

T = TypeVar("T")
ListOrTuple = TypeAliasType("ListOrTuple", Union[List[T], Tuple[T]], type_params=(T,))
ListOrTuple2: TypeAlias = Union[List[T], Tuple[T]]


def x(a: ListOrTuple[str], b: ListOrTuple2[str]) -> None:
    pass


x(a=["1"], b=["s"])
x(a=[1], b=["s"])
x(a=["s"], b=[1])

```

Mypy on both Python 3.8 and and 3.11 gives me the expected errors:

```bash
hatch shell hatch-test.py3.8   
mypy mypy_test.py  # <- assuming the above code is stored in such a file

hatch shell hatch-test.py3.11
mypy mypy_test.py
```
<img width="655" alt="image" src="https://github.com/user-attachments/assets/1d92054c-ac75-4c51-90e9-4a215997abdc">


Not saying that we should not use `TypeAliasType`, I'm just wondering when we want to use it and when we should use `TypeAlias` or if, for our purposes, it does not make a difference.

Reply:
Hmm, you're right @binste and I can't seem to find a clear answer on this I'm afraid. 
Quite odd that the doc gives an example that isn't motivating for usage.

I've been using `TypeAliasType` when any `TypeVar`(s) are involved. 
AFAIK, if we followed that as a convention we'd only need to update the definition for `Optional`.

https://github.com/vega/altair/blob/f542e9e8f1e4bf0fa4fe64e89dc5e72b94881cd2/tools/schemapi/schemapi.py#L804-L808

`TypeAlias` is technically [deprecated](https://docs.python.org/3/library/typing.html#typing.TypeAlias) as of `3.12` but until we have `3.12` as a lower bound - I would prefer using `TypeAlias` for simple aliases that are non-generic - as `TypeAliasType` is quite verbose.

Then once we reach `3.12` they can be replaced with `type` statements.

---

I did find that the `...` runtime issue I encountered is known:
- https://github.com/python/typing_extensions/issues/448
- https://github.com/python/typing_extensions/issues/48
- https://github.com/python/typing_extensions/issues/110

Reply:
I can't find something more definite either but was an interesting journey :) Thanks for all the comments! `TypeAlias` for simple and `TypeAliasType` for more complex syntax is ok with me